product_code/report_service/impl/src/main/scala/de/upb/cs/uc4/report/impl/ReportServiceImpl.scala

```diff
@@ -0,0 +1,191 @@
+        case Rejected(statusCode, reason) =>
+          log.error(s"Report of $username can't be persisted.", UC4Exception(statusCode, reason))
+      }
+    }.recover {
+      case exception: Exception => log.error(s"Report of $username can't be created.", exception)
+    }
+
+    Done
+  }
+
+  /** Get all data for the specified user */
+  def convertReadyReportToZip(username: String): Future[ByteString] =
+    entityRef(username).ask[ReportWrapper](replyTo => GetReport(replyTo)).map {
+      reportWrapper =>
+        val report = reportWrapper
+        val zipPath = Paths.get(System.getProperty("java.io.tmpdir"), s"report_$username", "report.zip")
+        val zipFile = new ZipFile(zipPath.toFile)
+
+        val folderPath = Paths.get(System.getProperty("java.io.tmpdir"), s"report_$username")
+        val reportPath = Paths.get(folderPath.toString, "userdata.json")
+
+        if (!folderPath.toFile.exists()) {
+          folderPath.toFile.mkdirs()
+        }
+
+        val reportTxt = Json.prettyPrint(Json.toJson(report))
+        var array: Array[Byte] = null
+
+        Using(new FileWriter(reportPath.toFile)) { writer =>
+          writer.write(reportTxt)
+        }
+
+        zipFile.addFile(reportPath.toFile)
+
+        array = Files.readAllBytes(zipFile.getFile.toPath)
+
+        val directory = new Directory(folderPath.toFile)
+        directory.deleteRecursively()
+
+        ByteString(array)
+    }
+
+  /** Get all data for the specified user */
+  override def getUserReport(username: String): ServiceCall[NotUsed, ByteString] = identifiedAuthenticated(AuthenticationRole.All: _*) {
+    (authUser, authRole) =>
+      ServerServiceCall { (header, _) =>
+        if (authUser != username.trim) {
+          throw UC4Exception.OwnerMismatch
+        }
+        val ref = entityRef(authUser)
+        ref.ask[ReportWrapper](replyTo => GetReport(replyTo)).flatMap { reportWrapper =>
+          reportWrapper.state match {
+            case ReportStateEnum.None =>
+              val timestamp = Calendar.getInstance().getTime.toString
+              ref.ask[Confirmation](replyTo => PrepareReport(timestamp, replyTo)).map {
+                case Accepted(_) =>
+                  prepareUserData(username, authRole, header, timestamp)
+                  (
+                    ResponseHeader(202, MessageProtocol.empty, List())
+                    .addHeader("X-UC4-Timestamp", timestamp),
+                    ByteString.empty
+                  )
+
+                case Rejected(statusCode, reason) => throw UC4Exception(statusCode, reason)
+              }
+            case ReportStateEnum.Preparing =>
+              Future.successful(
+                ResponseHeader(202, MessageProtocol.empty, List())
+                  .addHeader("X-UC4-Timestamp", reportWrapper.timestamp.get),
+                ByteString.empty
+              )
+            case ReportStateEnum.Ready =>
+              convertReadyReportToZip(username).map {
```

Comment:
Do we need additional error handling in case `convertReadyReportToZip` fails?

Reply:
Yes, we would throw a 500 at the moment

Reply:
A recover statement has been added, in order to properly log the error